core/src/backend/render.rs

```diff
@@ -213,25 +213,85 @@ pub enum JpegTagFormat {
 }
 
 /// Decoded bitmap data from an SWF tag.
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug)]
 pub struct Bitmap {
-    pub width: u32,
-    pub height: u32,
-    pub data: BitmapFormat,
+    width: u32,
+    height: u32,
+    format: BitmapFormat,
+    data: Vec<u8>,
 }
 
-/// Decoded bitmap data from an SWF tag.
-/// The image data will have pre-multiplied alpha.
-#[derive(Debug, Clone)]
-pub enum BitmapFormat {
-    Rgb(Vec<u8>),
-    Rgba(Vec<u8>),
+impl Bitmap {
+    /// Ensures that `data` is the correct size for the given `width` and `height`.
+    pub fn from_data(width: u32, height: u32, format: BitmapFormat, mut data: Vec<u8>) -> Self {
+        let expected_len = match format {
+            BitmapFormat::Rgb => width as usize * height as usize * 3,
+            BitmapFormat::Rgba => width as usize * height as usize * 4,
+        };
+        // If the size is incorrect, either we screwed up or the decoder screwed up.
+        debug_assert_eq!(data.len(), expected_len);
+        if data.len() != expected_len {
+            log::warn!(
+                "Incorrect bitmap data size, expected {} bytes, got {}",
+                data.len(),
+                expected_len
+            );
+            // Truncate or zero pad to the expected size.
+            data.resize(expected_len, 0);
+        }
+        Self {
+            width,
+            height,
+            format,
+            data,
+        }
+    }
+
+    pub fn to_rgba(mut self) -> Self {
```

Comment:
Maybe change this method to return just `Vec<u8>` instead of a whole `Bitmap`?
This will make the `assert!()` below redundant.

Reply:
I think returning a `Bitmap` is nicer in case anyone wants to store the result, although no one is doing so at the moment.